frontend/pre_analysis.py

```diff
@@ -0,0 +1,133 @@
+                    attributes.add(cls_stmt.name)
+                    if len(cls_stmt.args.args) == 0:
+                        continue
+                    first_arg = cls_stmt.args.args[0].arg  # In most cases it will be 'self'
+
+                    # Get attribute assignments where attribute value is the same as the first argument
+                    func_nodes = ast.walk(cls_stmt)
+                    assignments = [node for node in func_nodes if isinstance(node, ast.Assign)]
+
+                    for assignment in assignments:
+                        for target in assignment.targets:
+                            if (isinstance(target, ast.Attribute) and isinstance(target.value, ast.Name) and
+                                    target.value.id == first_arg):
+                                attributes.add(target.attr)
+                elif isinstance(cls_stmt, ast.Assign):
+                    # Get attributes defined as class-level assignment
+                    for target in cls_stmt.targets:
+                        if isinstance(target, ast.Name):
+                            attributes.add(target.id)
+
+        return class_to_attributes, class_to_base
+
+
+def propagate_attributes_to_subclasses(class_defs):
+    inheritance_forest = get_inheritance_forest(class_defs)
+    roots = get_forest_roots(inheritance_forest)
+    name_to_node = class_name_to_node(class_defs)
+
+    for root in roots:
+        propagate(root, inheritance_forest, name_to_node)
+
+
+def propagate(node, inheritance_forest, name_to_node):
+    for subclass in inheritance_forest[node]:
+        base_node = name_to_node[node]
+        sub_node = name_to_node[subclass]
+        sub_funcs_names = [func.name for func in sub_node.body if isinstance(func, ast.FunctionDef)]
+
+        inherited_funcs = [func for func in base_node.body
+                           if isinstance(func, ast.FunctionDef) and func.name not in sub_funcs_names]
+        sub_node.body += inherited_funcs
+        propagate(subclass, inheritance_forest, name_to_node)
+
+
+def class_name_to_node(nodes):
+    name_to_node = {}
+    for node in nodes:
+        name_to_node[node.name] = node
+    return name_to_node
+
+
+def get_forest_roots(forest):
+    roots = list(forest.keys())
+    for node in forest:
+        for sub in forest[node]:
+            if sub in roots:
+                roots.remove(sub)
+    return roots
+
+
+def get_inheritance_forest(class_defs):
+    tree = {}
+    for cls in class_defs:
+        tree[cls.name] = []
+    for cls in class_defs:
+        bases = cls.bases
+        for base in bases:
+            tree[base.id].append(cls.name)
+    return tree
+
+
+def add_init_if_not_existing(class_node):
+    for stmt in class_node.body:
+        if isinstance(stmt, ast.FunctionDef) and stmt.name == "__init__":
+            return
+    class_node.body.append(ast.FunctionDef(
```

Comment:
- For **1**, class instantiation is handled by treating it as a call to an `__init__` function, so I thought it is much easier to generate it in the pre-analysis than to handle such special case in Z3.
- For **2**, this behavior will not happen because the methods propagation happens before the `init` addition, so B will inherit the correct `__init__` from A and **no** new `__init__` function will be generated.